### PR TITLE
Add default behavior for driving, non driving and sidewalks

### PR DESCRIPTION
**Pull request by tnederlof**
I took a stab at building more sensible default behavior. I am very new to rust so sorry if it's not very idiomatic, I figure the only way to learn is to start writing and get critiqued!

This works pretty well for streets regardless of the side of the street is Fwd and Back (it looks at the number of Fwd and Back lanes, defaults to Fwd unless there are too many Fwd in which case will make the direction Back). Currently it looks at all of the lanes to determine this balance (I could also see the argument to only look at drivable lanes to determine the balance of Fwd and Back).

For other types of lanes, there are a lot of edge cases as I looked around the map. If for example there is a parked car lane on the side where a bike lane gets added, the bike lane will be sandwiched between the sidewalk and the parked lane. I think this is okay as this is just a default that can be tweaked. Sometimes one-way streets will cause issues too where if there are two drive lanes Fwd then an added sidewalk will have the direction set to Back which probably doesn't make sense. Maybe sidewalks should always be set in the direction of the nearest lane?

### CODE REVIEW

File: game/src/edit/roads.rs [hunk 1]
```diff
@@ -508,8 +508,34 @@ fn can_reverse(_: LaneType) -> bool {
...
+            0
+        }
+    } else {
+        if road.lanes_ltr.last().unwrap().lt.is_walkable() {
+            road.lanes_ltr.len() - 1
+        } else {
+            road.lanes_ltr.len()
+        }
+    };
+    idx
```
Comment (dabreegster) [inline game/src/edit/roads.rs:525]: Style: You could just make the entire function be an expression, aka, take away the `let idx = ` part

File: game/src/edit/roads.rs [hunk 2]
```diff
@@ -508,8 +508,34 @@ fn can_reverse(_: LaneType) -> bool {
     lt == LaneType::Driving || lt == LaneType::Biking || lt == LaneType::Bus
 }*/
 
+fn default_outside_lane_placement(road: &mut EditRoad, dir: Direction) -> usize {
+    let idx = if road.lanes_ltr.first().unwrap().dir == dir {
```
Comment (dabreegster) [inline game/src/edit/roads.rs:512]: Style: Instead of `.first().unwrap()`, we can just do `road.lanes_ltr[0].dir`. It's equivalent in terms of crashing if the vector is empty; `first` returns `Option<...>` and `unwrap` panics if it's `None`.

To note, it's guaranteed that `road.lanes_ltr` is always non-empty, so that's why we can do `.last().unwrap()` and not worry about an error case anywhere here.

File: game/src/edit/roads.rs [hunk 3]
```diff
@@ -508,8 +508,34 @@ fn can_reverse(_: LaneType) -> bool {
...
+        if road.lanes_ltr.first().unwrap().lt.is_walkable() {
+            1
```
Comment (dabreegster) [inline game/src/edit/roads.rs:514]: Logic: Just talking myself through things here: what about the case of a one-way when we have

sdds
v^^^

(backwards sidewalk, fwds driving, fwds driving, fwds sidewalk)

That would insert a backwards driving lane at idx 1. Which is still the reasonable spot, so cool.

File: game/src/edit/roads.rs [hunk 4]
```diff
@@ -508,8 +508,34 @@ fn can_reverse(_: LaneType) -> bool {
...
+    };
+    idx
+}
+
 // Returns the index where the new lane was inserted
 fn add_new_lane(road: &mut EditRoad, lt: LaneType) -> usize {
+
+    // Default to Fwd as long as the lane directions are equal or there are more with Back
+    // if there are more lanes with Fwd then use Back
+    let dir = if (road.lanes_ltr.iter().filter(|x| x.dir == Direction::Fwd).count() as f64 / road.lanes_ltr.len() as f64) <= 0.5 {
```
Comment (dabreegster) [inline game/src/edit/roads.rs:533]: You can auto-format the code and get better line-wrapping here by running `cargo +nightly fmt`. You have to install the nightly toolchain for the formatting, unfortunately, because we [use a few overrides](https://github.com/a-b-street/abstreet/blob/master/rustfmt.toml) that aren't stable yet.

File: game/src/edit/roads.rs [hunk 5]
```diff
@@ -508,8 +508,34 @@ fn can_reverse(_: LaneType) -> bool {
...
+            road.lanes_ltr.len()
+        }
+    };
+    idx
+}
+
 // Returns the index where the new lane was inserted
 fn add_new_lane(road: &mut EditRoad, lt: LaneType) -> usize {
+
+    // Default to Fwd as long as the lane directions are equal or there are more with Back
```
Comment (dabreegster) [inline game/src/edit/roads.rs:531]: Style: I got slightly confused reading this, how about `// If there are more lanes pointing forwards, then insert the new one backwards, and vice versa`

File: game/src/edit/roads.rs [hunk 6]
```diff
@@ -518,29 +544,25 @@ fn add_new_lane(road: &mut EditRoad, lt: LaneType) -> usize {
...
         LaneType::Biking | LaneType::Bus | LaneType::Parking | LaneType::Construction => {
-            if road.lanes_ltr.last().unwrap().lt.is_walkable() {
-                road.lanes_ltr.len() - 1
-            } else {
-                road.lanes_ltr.len()
-            }
+            default_outside_lane_placement(road, dir)
         }
-        // Place it where it's missing
+        // Place it where it's missing, and if both sides have or do not have sidewalks then go with the dir side
```
Comment (dabreegster) [inline game/src/edit/roads.rs:551]: The new road editor doesn't enforce this yet, but sidewalks are only supposed to be at the beginning or end, never in the middle. So I think we can actually keep the previous logic here, since it's a bit simpler. If the road has 0 sidewalks, I don't think it's important to initially place the first one on one side or the other.

I've yet to enforce the "sidewalks only on the outside" rule, because I'm not sure how the "move left/right" control should work for sidewalks. Maybe since they're kind of special, the UI should even just have two checkboxes, "sidewalk on the left" & "sidewalk on the right". Except we do still want to let people set the width... so maybe they're not so special after all. Not sure yet.

**Review by dabreegster — COMMENTED** [review]
Quite a solid start! Just playing around with this, it feels much more natural to have new lanes balance out the two directions.

> Currently it looks at all of the lanes to determine this balance (I could also see the argument to only look at drivable lanes to determine the balance of Fwd and Back).

Maybe we could even just filter by the type of lane someone's trying to add -- if there's already a bus lane on one side, adding another bus lane should default to the other, even if it already has more lanes total.

> If for example there is a parked car lane on the side where a bike lane gets added, the bike lane will be sandwiched between the sidewalk and the parked lane.

There's probably a regional default here that we could be smart about -- just in Seattle, I've seen both styles, where parking is used as a buffer for the bike lane, and where the bike lane is in the "door zone" between vehicle traffic and parking. The second is probably more common here.

> Maybe sidewalks should always be set in the direction of the nearest lane?

Maybe something like this... sidewalks and parking lanes have some special rules for direction that I can't quite summarize/express. But when it's wrong, you can see it:
Screenshot from 2021-04-29 17-31-50
Because the sidewalk on the top is Fwd, the "SharedSidewalkCorner" rendered in the intersection happens instead of a crosswalk. And since the parking lane on the top is Fwd, the "T"-hatches point the wrong way.


Stepping back, we could think more systematically about testing edge cases. There are different factors that tend to complicate things:

- one-way roads
- left-handed driving
- contraflow bus/bike lanes on a one-way
- two-way cycle tracks on a two-way road (so the Fwd/Back direction actually flips 3 times

We could put together unit tests that specify the starting conditions for the road and what someone's trying to insert, then assert that the ending conditions match whatever we decide is intuitive. I'd be fine with that, but as you may have noticed, there's not too many tests in the codebase. I'm fine continuing without them here, especially since we're not even sure about the best behavior for some of these edge cases.

File: game/src/edit/roads.rs (hunk 4 above)
Comment (tnederlof) [reply game/src/edit/roads.rs:533]: Very cool I will check it out

File: game/src/edit/roads.rs (hunk 6 above)
Comment (tnederlof) [reply game/src/edit/roads.rs:551]: Agreed, I will simplify it here.

**Comment by dabreegster** [comment]
> Honestly I would love to learn some testing in rust at some point

https://doc.rust-lang.org/beta/rust-by-example/testing/unit_testing.html is a good intro, and https://github.com/a-b-street/abstreet/blob/805160874765f7e7b7b6af3435b873f55667a5a6/map_model/src/make/initial/lane_specs.rs#L383 is an example of unit testing the function that produces `LaneSpecs` from the original OSM tags. I wrote Go before coming to Rust, and the "table-driven test" pattern seems well-suited here too -- a bunch of (input, expected output) bundles, along with helper functions for showing the diff sanely.

**Comment by tnederlof** [comment]
I rewrote some of the logic trying to balance sensible defaults without writing to many specific cases.

Driving now looks at existing driving lanes to always defaults to either forward if equal or if an imbalance goes with the minority to balance the road.

For bike/bus/parked/construction lanes in the absence of any on the road it will default to the forward side of the road and if there is an existing of that lane type it will balance it out by placing it on the other side. The lanes are always placed on the outside or in the case of a walkable lane the next index inwards. This will place bike lanes between a sidewalk and parked cars but that might be an okay default and makes the code similar (don't need to look at what parked lanes currently exist, etc.).

For sidewalks I returned the index code to the prior code to simplify. The only add is setting the direction of the new sidewalk based on the next lane inwards direction (which I think is a decent default but might be some edge cases).

There are still some edge cases with one way roads, weird lane configurations, etc. Tests will help with that.

**Comment by dabreegster** [comment]
The behavior feels natural in some ad-hoc tests in US and UK maps -- really nice! And the code style looks fine, so I'd say this is ready to merge. Want to continue with edge cases and tests in this PR, or merge this one and start another later? Either's fine by me, though I might merge this one before the Sunday release, since it's such a big improvement to the new editor.

**Comment by tnederlof** [comment]
> 
> 
> The behavior feels natural in some ad-hoc tests in US and UK maps -- really nice! And the code style looks fine, so I'd say this is ready to merge. Want to continue with edge cases and tests in this PR, or merge this one and start another later? Either's fine by me, though I might merge this one before the Sunday release, since it's such a big improvement to the new editor.

Glad you like it. Yes, I ran the formatter with the nightly build, thanks for letting me know, autoformatter is really helpful especially being new as I dont know the right ways to format to begin with.

I am cool with it being merged now so its generally useful for most cases, shouldnt be any more messed up than before on some of the edge cases anyway. Maybe I will file an issue about the tests so I remember to come back and do the tests sometime? I might take a break from this and look at the 15min city and/or other things for a bit and circle back. Also learning rust more generally, its a cool language so far!

**Comment by dabreegster** [comment]
Thanks for the PR!
